src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1861,156 +1819,155 @@ final void registerWorker(WorkQueue w) {
      * @param ex the exception causing failure, or null if none
      */
     final void deregisterWorker(ForkJoinWorkerThread wt, Throwable ex) {
+        if ((runState & STOP) != 0L)       // ensure released
+            releaseAll();
         WorkQueue w = null;
         int src = 0, phase = 0;
         boolean replaceable = false;
         if (wt != null && (w = wt.workQueue) != null) {
             phase = w.phase;
-            if ((src = w.source) != DEREGISTERED) { // else trimmed on timeout
-                w.source = DEREGISTERED;
-                if (phase != 0) {         // else failed to start
+            if ((src = w.source) != DROPPED) {
+                w.source = DROPPED;        // else already dropped
```

Comment:
@DougLea Would there be any gain in atomically swapping out `w.source` to `DROPPED`?

Reply:
No the source field is only written by owner. (With one harmless exception -- cascading idle timeouts, where an invalid value is set to check that no tasks were taken by successor.) I should note this somewhere.